src/rpc/methods/eth.rs

```diff
@@ -1955,54 +1955,81 @@ impl RpcMethod<1> for EthGetTransactionByHash {
         ctx: Ctx<impl Blockstore + Send + Sync + 'static>,
         (tx_hash,): Self::Params,
     ) -> Result<Self::Ok, ServerError> {
-        let message_cid = ctx.chain_store().get_mapping(&tx_hash)?.unwrap_or_else(|| {
-            tracing::debug!(
-                "could not find transaction hash {} in Ethereum mapping",
-                tx_hash
-            );
-            // This isn't an eth transaction we have the mapping for, so let's look it up as a filecoin message
-            tx_hash.to_cid()
-        });
+        get_eth_transaction_by_hash(ctx, tx_hash, None).await
+    }
+}
 
-        // First, try to get the cid from mined transactions
-        if let Ok(Some((tipset, receipt))) = ctx
-            .state_manager
-            .search_for_message(None, message_cid, None, Some(true))
-            .await
-        {
-            let ipld = receipt.return_data().deserialize().unwrap_or(Ipld::Null);
-            let message_lookup = MessageLookup {
-                receipt,
-                tipset: tipset.key().clone(),
-                height: tipset.epoch(),
-                message: message_cid,
-                return_dec: ipld,
-            };
-
-            if let Ok(tx) = new_eth_tx_from_message_lookup(&ctx, &message_lookup, None) {
-                return Ok(Some(tx));
-            }
-        }
+pub enum EthGetTransactionByHashLimited {}
+impl RpcMethod<2> for EthGetTransactionByHashLimited {
+    const NAME: &'static str = "Filecoin.EthGetTransactionByHashLimited";
+    const NAME_ALIAS: Option<&'static str> = Some("eth_getTransactionByHashLimited");
+    const PARAM_NAMES: [&'static str; 2] = ["tx_hash", "limit"];
+    const API_PATHS: ApiPaths = ApiPaths::V1;
+    const PERMISSION: Permission = Permission::Read;
 
-        // If not found, try to get it from the mempool
-        let (pending, _) = ctx.mpool.pending()?;
-
-        if let Some(smsg) = pending.iter().find(|item| item.cid() == message_cid) {
-            // We only return pending eth-account messages because we can't guarantee
-            // that the from/to addresses of other messages are conversable to 0x-style
-            // addresses. So we just ignore them.
-            //
-            // This should be "fine" as anyone using an "Ethereum-centric" block
-            // explorer shouldn't care about seeing pending messages from native
-            // accounts.
-            if let Ok(eth_tx) = EthTx::from_signed_message(ctx.chain_config().eth_chain_id, smsg) {
-                return Ok(Some(eth_tx.into()));
-            }
+    type Params = (EthHash, Option<ChainEpoch>);
```

Comment:
Why did you choose an`Option<ChainEpoch>` and not just a `ChainEpoch`?

Reply:
because `look_back_limit` is of type `Option<i64>` in `search_for_message`
```
forest_filecoin::state_manager::StateManager
impl<DB> StateManager<DB>
pub async fn search_for_message(self: &Arc<Self>, from: Option<Arc<Tipset>>, msg_cid: Cid, look_back_limit: Option<i64>, allow_replaced: Option<bool>) -> Result<Option<(Arc<Tipset>, Receipt)>, Error>
where
    // Bounds from impl:
    DB: Blockstore + Send + Sync + 'static,
```

Reply:
Then let's add another test that checks the case where chain epoch is `None`.

Reply:
https://github.com/ChainSafe/forest/pull/4774/files#r1766769807

Reply:
You're speaking of testing `get_eth_transaction_by_hash` function which I think is not the same.

Adding the test:

```
            tests.push(RpcTest::identity(EthGetTransactionByHashLimited::request(
                (tx.hash.clone(), None),
            )?));
```

results in:

```
./forest-tool api compare --lotus /ip4/127.0.0.1/tcp/1234/http --forest /ip4/127.0.0.1/tcp/2345/http forest_snapshot_calibnet_2024-09-10_height_1954295.forest.car.zst --filter EthGetTransactionByHashLimited
| RPC Method                                  | Forest            | Lotus |
|---------------------------------------------|-------------------|-------|
| Filecoin.EthGetTransactionByHashLimited (4) | CustomCheckFailed | Valid |
| Filecoin.EthGetTransactionByHashLimited (8) | Valid             | Valid |
Error: Some tests failed
```

Reply:
changed `limit` type to `ChainEpoch`